docs: editorial love for IBC overview (#544)
cosmos/ibc-go: docs/ibc/overview.md

Since IBC works over a distributed network and relies on potentially faulty relayers to relay messages between ledgers, 
IBC must handle the case where a packet does not get sent to its destination in a timely manner or at all. Thus, packets must 
specify a timeout height or timeout timestamp, after which a packet can no longer be successfully received on the destination chain.

If the timeout does get reached, then a proof of packet timeout can be submitted to the original chain, which can then perform 
application-specific logic to timeout the packet, perhaps by rolling back the packet send changes (refunding senders any locked funds, etc.).

- In ORDERED channels, a timeout of a single packet in the channel causes the channel to close. If packet sequence `n` times out, 
then no packet at sequence `k > n` can be successfully received without violating the contract of ORDERED channels that packets are processed in the order that they are sent. Since ORDERED channels enforce this invariant, a proof that sequence `n` hasn't been received on the destination chain by the specified timeout of packet `n` is sufficient to timeout packet `n` and close the channel.

> - In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, one must provide a proof that a packet receipt does not exist for the packet's sequence by the specified timeout. Of course, timing out a packet on an UNORDERED channel simply triggers the application-specific timeout logic for that packet, and does not close the channel.

Author: I didn't tackle this grammatical challenge, but now that we are all focused on clarity and overall improvements, let's see if we can't improve here 

```suggestion
- In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, proof is required that a packet receipt does not exist for the packet's sequence by the specified timeout. Of course, timing out a packet on an UNORDERED channel simply triggers the application-specific timeout logic for that packet and does not close the channel.
```

> - In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, one must provide a proof that a packet receipt does not exist for the packet's sequence by the specified timeout. Of course, timing out a packet on an UNORDERED channel simply triggers the application-specific timeout logic for that packet, and does not close the channel.

Reviewer: ```suggestion
- In UNORDERED channels, packets can be received in any order. Thus, IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information; it is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. In order to timeout a packet on an UNORDERED channel, a proof is required that a packet receipt **does not exist** for the packet's sequence by the specified timeout deadline. Timing out a packet on an UNORDERED channel prevents any further receiving of the packet, triggers the application-specific timeout logic for that packet, but does not close the channel.
```

Author: see how the rewrite is now... let's try to write concisely what is happening. Instead of "deadline" let's use "timeout" (since it's timeout height, not a date-based deadline)

let's say what happens, not what doesn't happen

Very challenging! I feel for our users

Author: oh man, now I went and rewrote a bit more for clarity

For this reason, most modules should use UNORDERED channels as they require less liveness guarantees to function effectively for users of that channel.

### [Acknowledgments](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel)

Modules can also choose to write application-specific acknowledgments upon processing a packet. This acknowledgment can be done synchronously on `OnRecvPacket` if the module processes packets as soon as they are received from IBC module. Or acknowledgments can be done asynchronously if module processes packets at some later point after receiving the packet.

> Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. This should be decided through version negotiation during the channel handshake.

Author: ```suggestion
Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. Define this encoding through version negotiation during the channel handshake.
```

Author: I probably guessed incorrectly but again the unclear reference "This" in "This should be decided"

are we talking about the encoding? or ? and how is it decided?

Reviewer 2: I think we're talking about how it is decided the encoding, so that both parties know how to encode/decode to understand each other.

Reviewer: `The encoding should be negotiated between the two parties during version negotiation in the channel handshake.

Author: since this is the overview topic, I suggest that we are too specific here.

Readers can learn more at https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation 
---
oh "should" is such an ambiguous word, let's avoid per this advice
https://developers.google.com/style/word-list?hl=en#letter-s 

**should, should be
Generally avoid.**
![image](https://user-images.githubusercontent.com/10067215/142517735-4665c117-9deb-46ae-a413-2aede50b63ea.png)

> Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. This should be decided through version negotiation during the channel handshake.

Reviewer: ```suggestion
Regardless, this acknowledgment data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. It is incumbent on receiver modules to encode their acknowledgment so that the sender module can decode it correctly. The encoding should be negotiated between the two parties during version negotiation in the channel handshake.
```

Author: yes, much better @AdityaSripal (except "should" let's use "must" instead)

The acknowledgment can encode whether the packet processing succeeded or failed, along with additional information that allows the sender module to take appropriate action.

> After the acknowledgment has been written by the receiving chain, a relayer relays the acknowledgment back to the original sender module which  then executes application-specific acknowledgment logic using the contents of the acknowledgment. This can involve rolling back packet-send changes in the case of a failed acknowledgment (refunding senders).

Author: The unclear reference (the notorious naked "This") here makes me wonder what the subject is (I couldn't even guess... "This logic"? or "This relay"?)

please r/w or provide the subject of "This" in  

"This can involve rolling"

Reviewer 2: I think it refers to the execution of the acknowledgment logic. So maybe something like `This acknowledgment logic can involve rolling...`.

There is also an extra space in `which  then`.

Reviewer: Maybe we should replace that sentence with:

`Application developers can roll back packet-send changes in the case of a failed acknowledgement (e.g. refunding senders in ICS20).`

Author: yes better, the more I read this content, the move love it needs!

> After the acknowledgment has been written by the receiving chain, a relayer relays the acknowledgment back to the original sender module which  then executes application-specific acknowledgment logic using the contents of the acknowledgment. This can involve rolling back packet-send changes in the case of a failed acknowledgment (refunding senders).

Reviewer: ```suggestion
After the acknowledgment has been written by the receiving chain, a relayer relays the acknowledgment back to the original sender module which  then executes application-specific acknowledgment logic using the contents of the acknowledgment. Application developers can roll back packet-send changes in the case of a failed acknowledgement (e.g. refunding senders in ICS20).
```

chain B using an already established connection:

1. chain A sends a `ChanOpenInit` message to signal a channel initialization attempt with chain B.
2. chain B sends a `ChanOpenTry` message to try opening the channel on chain A.
3. chain A sends a `ChanOpenAck` message to mark its channel end status as open.
4. chain B sends a `ChanOpenConfirm` message to mark its channel end status as open.

If all handshake steps are successful, the channel is opened on both sides. At each step in the handshake, the module
associated with the `ChannelEnd` executes its callback. So
on `ChanOpenInit`, the module on chain A executes its callback `OnChanOpenInit`.

Just as ports came with dynamic capabilities, channel initialization returns a dynamic capability
that the module **must** claim so that they can pass in a capability to authenticate channel actions
like sending packets. The channel capability is passed into the callback on the first parts of the
handshake; either `OnChanOpenInit` on the initializing chain or `OnChanOpenTry` on the other chain.

### [Packets](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel)

> Modules communicate with each other by sending packets over IBC channels. All

Author: is there a reason we don't use the packet definition from https://github.com/cosmos/ibc/blob/master/spec/core/ics-004-channel-and-packet-semantics/README.md 

(seems much more clear!)

Reviewer: I'm in favor of just copying that over.

The generic `Identifier` type in the spec should be changed to `string` for the go implementation

Author: @AdityaSripal can you take on that spec work?

IBC packets contain the destination `portID` and `channelID` along with the source `portID` and
`channelID`. This packet structure allows modules to know the sender module of a given packet. IBC packets 

> contain a sequence to optionally enforce ordering. 

Author: so the packet sequence that defines ordering is 
  `sequence: uint64` ?
  seems like we would want to define what is ORDERED and what is UNORDERED

Reviewer: Yes if a channel is ORDERED, then the packets must be received in the order of the `sequence` that is inside the packet.

Thus, packet data is opaque to IBC handlers. It is incumbent on a sender module to encode
their application-specific packet information into the `Data` field of packets. The receiver
module must decode that `Data` back to the original application data.

### [Receipts and Timeouts](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel)

Since IBC works over a distributed network and relies on potentially faulty relayers to relay messages between ledgers, 
IBC must handle the case where a packet does not get sent to its destination in a timely manner or at all. Packets must 
specify a non-zero value for timeout height (`TimeoutHeight`) or timeout timestamp (`TimeoutTimestamp` ) after which a packet can no longer be successfully received on the destination chain.

- The `timeoutHeight` indicates a consensus height on the destination chain after which the packet is no longer be processed, and instead counts as having timed-out.
- The `timeoutTimestamp` indicates a timestamp on the destination chain after which the packet is no longer be processed, and instead counts as having timed-out.

> If the timeout passes without the packet being successfully received, the
> sending module can timeout the packet and take appropriate actions.

Reviewer: ```suggestion
If the timeout passes without the packet being successfully received, the packet can no longer be
received on the destination chain. The sending module can timeout the packet and take appropriate actions.
```

If the timeout is reached, then a proof of packet timeout can be submitted to the original chain. The original chain can then perform 
application-specific logic to timeout the packet, perhaps by rolling back the packet send changes (refunding senders any locked funds, etc.).

> - In ORDERED channels, a timeout of a single packet in the channel causes the channel to close. 

Author: since line 175 defines what happens for a packet timeout in an ORDERED channel, I attempted to describe what happens in an UNORDERED channel  on line 180

(we really gotta show this essential content some love... am I on the right track here?)

## What is the Interblockchain Communication Protocol (IBC)?

This document serves as a guide for developers who want to write their own Inter-Blockchain
Communication protocol (IBC) applications for custom use cases.

Due to the modular design of the IBC protocol, IBC
application developers do not need to be concerned with the low-level details of clients,
connections, and proof verification. 

This brief explanation of the lower levels of the
stack is provided to give application developers a high-level understanding of the IBC

> protocol. Details on the abstraction layer are most relevant for application
> developers (channels and ports) and describe how to define custom packets and

Reviewer: ```suggestion
protocol. The channel and port abstraction layer is most relevant to application developers, here we will define custom packets and
```

The second sentence didn't make sense to me.

`IBCModule` callbacks.

To have your module interact over IBC you must: 

- Bind to a port or ports
- Define your own packet data

> - Define optional acknowledgement structs

Author: I wasn't sure what is optional, the acknowledgement structs or defining them

Reviewer: You can define your own custom acknowledgement struct if you wish, or you can use the default one provided by core IBC.

Author: ```suggestion
- Use the default acknowledgment struct provided by core IBC, or optionally define your own custom acknowledgment struct
```

> - Know how to encode and decode the packet data

Reviewer: ```suggestion
- Standardize an encoding of the packet data
```

Reviewer: or you could say `Define` to keep it consistent

Reviewer: Though this point seems like it should be rolled into `Defining packet data struct`.

Perhaps `Define packet data struct and encoding`?

The same is true for acknowledgment

> - Implement the `IBCModule` interface

Author: @crodriguezvega in this case, we can apply the Google developer documentation technical content style standards and add a period (.) at the end of each list item since the items include a verb.

My preference is to omit punctuation at the end of sentence fragments in lists; the key point is to be consistent throughout the content set. 

https://developers.google.com/style/periods?hl=en#bulleted-lists

client's consensus state. A client can be associated with any number of connections to the counterparty
chain. 

The supported IBC clients are:

* [Solo Machine light client](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/06-solomachine): Devices such as phones, browsers, or laptops.
* [Tendermint light client](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint): The default for Cosmos SDK-based chains.
* [Localhost (loopback) client](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/09-localhost): Useful for
testing, simulation, and relaying packets to modules on the same application.

### [Connections](https://github.com/cosmos/ibc-go/blob/main/modules/core/03-connection)

Connections encapsulate two `ConnectionEnd` objects on two separate blockchains. Each
`ConnectionEnd` is associated with a client of the other blockchain (for example, the counterparty blockchain).
The connection handshake is responsible for verifying that the light clients on each chain are
correct for their respective counterparties. Connections, once established, are responsible for

> facilitation all cross-chain verification of IBC state. A connection can be associated with any

Author: I changed all instances of may to can
per Google developer documentation style guide advice for using can https://developers.google.com/style/word-list#letter-c 
and  may https://developers.google.com/style/word-list#may

key point

> may
> In general, reserve _may_ for official policy or legal considerations. To convey possibility, use _can_ or _might_ instead. To convey permission, use _can_ instead. See also can, might, and should.

Reviewer 3: Should this not read as: "Connections, once established, are responsible for **facilitating** all cross-chain verification of IBC state"?

Another small point which I am unsure of but I assumed previously - I thought the use of `may` here was as per [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119). @AdityaSripal maybe can you confirm?

Reviewer: > Connections, once established, are responsible for facilitating all cross-chain verification of IBC state

Yea that's right.

`MAY` implied optional as mentioned in the RFC, but i don't believe we stuck to a very strict technical writing standard so I defer to experts on the best way to convey this.

Author: I'll use the gerund verb facilitating as recommended, good catch @damiannolan but then we need plural "verifications"

number of channels.

### [Proofs](https://github.com/cosmos/ibc-go/blob/main/modules/core/23-commitment) and [Paths](https://github.com/cosmos/ibc-go/blob/main/modules/core/24-host)
  
In IBC, blockchains do not directly pass messages to each other over the network. Instead, to

> communicate, a blockchain commits some state to a specifically defined path that is reserved for a

Author: I applied the guideline to use present tense, per https://developers.google.com/style/tense

specific message type and a specific counterparty. For example, for storing a specific connectionEnd as part
of a handshake, or a packet intended to be relayed to a module on the counterparty chain. A relayer
process monitors for updates to these paths and relays messages by submitting the data stored
under the path along with a proof to the counterparty chain. 

- The paths that all IBC implementations must use for committing IBC messages is defined in
[ICS-24 Host State Machine Requirements](https://github.com/cosmos/ics/tree/master/spec/core/ics-024-host-requirements). 
- The proof format that all implementations must be able to produce and verify is defined in [ICS-23 Proofs](https://github.com/confio/ics23) implementation.

### [Capabilities](https://github.com/cosmos/cosmos-sdk/blob/master/docs/core/ocap.md)

IBC is intended to work in execution environements where modules do not necessarily trust each
other. Thus, IBC must authenticate module actions on ports and channels so that only modules with the

> appropriate permissions can use them. This module action authentication is accomplished using a [dynamic

Author: for "This" I think we mean "This module action authentication"

beware the naked "This" especially for complex sentences

> this, that
> Where possible, put a noun after this or that for clarity. If doing so results in clunky prose, then don't do it; but even then, try thinking about what the noun would be. If you aren't sure what noun this or that refers to, then consider rephrasing.

see https://developers.google.com/style/word-list#letter-t

> appropriate permissions can use them. This module action authentication is accomplished using a [dynamic

Reviewer: ```suggestion
appropriate permissions can use them. This module authentication is accomplished using a [dynamic
```

Seems like `module authentication` is more appropriate

capability store](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-003-dynamic-capability-store.md). Upon binding to a port or
creating a channel for a module, IBC returns a dynamic capability that the module must claim in

> order to use that port or channel. This dynamic capability store prevents other modules from using that port or channel since

Author: another unclear reference "This"
I guessed that we are talking about "This dynamic capability store" 

But please teach me if I'm off track

> order to use that port or channel. This dynamic capability store prevents other modules from using that port or channel since

Reviewer: ```suggestion
order to use that port or channel. The dynamic capability module prevents other modules from using that port or channel since
```

handler. This requirement prevents a malicious module from opening channels with ports it does not own. Thus,
IBC modules are responsible for claiming the capability that is returned on `BindPort`.

### [Channels](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel)

An IBC channel can be established between two IBC ports. Currently, a port is exclusively owned by a
single module. IBC packets are sent over channels. Just as IP packets contain the destination IP
address and IP port as well as the source IP address and source IP port, IBC packets contain
the destination portID and channelID as well as the source portID and channelID. This packet structure enables IBC to
correctly route packets to the destination module while also allowing modules receiving packets to
know the sender module.

A channel can be `ORDERED`, in which case, packets from a sending module must be processed by the
receiving module in the order they were sent. Or a channel can be `UNORDERED`, in which case packets
from a sending module are processed in the order they arrive (might be a different order than they were sent).

> Modules can choose which channels they wish to communicate over with, thus IBC expects modules to

Reviewer: ```suggestion
Modules can choose which channels they wish to communicate over, thus IBC expects modules to
```
